drivers: ethernet: smsc9118 driver (as used by mps2_an385)

pfalcon opened the pull request:
This is initial WIP on implementing driver for SMSC9220/SMSC9118/LAN9118 Ethernet chips, as available in ARM MPS2 board (BOARD=mps2_an385). The main purpose of this driver is to be used with QEMU emulation however, it's not tested on a real hardware (so less attention is paid to matters like cable detection, etc.)

NOTE to reviewers: as it stands now, this is very dirty code, you probably want to skip looking at it for now. This is posted to reference it in other tickets and discuss ways to brush up the code with @galak, the primary requester of this driver.

pfalcon committed:
boards: arm: mps2_an385: Add DTS node for Ethernet controller

The board has SMSC LAN9220 (actually as an "IP core" in an
FPGA-emulated SoC). The patch includes DTS bindings for this device.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
@galak: So, this now basically works, at least on the level of pings. There's a lot to do both re: functionality, and of course re: cleanup. I'd like to discuss high-level matter regarding the latter:

1. Use of code imported from mbed. The API of the C file (smsc9220_eth.c) is more or less adhoc, I already work it around. I would plan to get rid of it, and access hardware registers directly in the driver. It thus would be more like eth_enc28j60.c than those which use HALs. The header would stay, but would be patched.
2. Naming. The chips are apparently due to the company "SMSC" and where called SMSC9220, etc. It apparently was bought by Microchip, which renamed them to LAN9220, etc. Also, while MPS2 uses SMSC9220, QEMU actually emulated SMSC9118. The difference is in the things like checksum offloading, etc., something which we won't use. So I guess, it's more faithful to call the driver eth_smsc9118.c. For reference, in Linux it's called smsc911x.c IIRC.

@galak, so please let me know if you see something terribly wrong with the above, otherwise I'll gradually move it in that direction. Thanks.

pfalcon commented:
Ok, so I decided to settle on naming it eth_smsc911x.c after all.

Functionality progresses well, pings now seem to be pretty reliable.

pfalcon commented:
Updated to @tbursztyka's recently merged refactors. Pings work.

pfalcon committed:
boards: mps2_an385: Select QEMU_TARGET

Board can be emulated in QEMU, so select QEMU_TARGET as required for
various bits of "make run" magic to work.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
samples: net: echo_server: Add overlay for eth_smsc911x

As can be used for example with BOARD=mps2_an385.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Update with qemu eth support, a-la https://github.com/zephyrproject-rtos/zephyr/pull/11138.

Now 
~~~
zephyr-make BOARD=mps2_an385 CONF_FILE="prj.conf overlay-smsc911x.conf" run
~~~

works.

pfalcon committed:
boards: mps2_an385: Enable SLIP networking

mps2_an385 is enabled for networking -> default networking is SLIP ->
SLIP selects UART_PIPE -> UART_PIPE requires UART_PIPE_ON_DEV defined
-> undefined leads to error on building some samples.

Fix all of these issues.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
So, this now passes CI. There's still more testing to do, after which more code cleanup to do. It also depends on other patches which were submitted separately, e.g. https://github.com/zephyrproject-rtos/zephyr/pull/12230 . 

The biggest decision point to make is what to do about former "smsc9220_eth.h". The situation with it is:

1. It comes from an older version of mbedOS (no longer in mbedOS master), and was quick-committed by @galak.
2. Its original name is "smsc9220_eth.h", but we need it as "eth_smsc911x_priv.h" (to follow existing naming convention in Zephyr). Of course, we don't want to have stray rename in Zephyr history, so I squashed the renamed back into the original @galak's commit.
3. Likewise, its code style is widely different from Zephyr code style, leading to checkpatch errors. So again, I made changes, and squashed them back into @galak's commit, to avoid noise in Zephyr history.
4. But further Zephyr-specific "semantic" changes, I keep in a separate commit.
5. As it stands now, eth_smsc911x_priv.h doesn't lead to checkpatch errors, but has quite a bunch of warnings.
6. One way to deal with all that would be to squash all eth_smsc911x_priv.h related commits down to one. As mentioned, it's no longer used in mbedOS, so preserving any similarity to original mbedOS shape is of little use. Otherwise, there's ARM copyright and a comment of origin.

@galak, let me know if you agree with the proposal in p.6 above. And let me know if you care if that commit is under your authorship or care it to be now (it would be very different from what *you* committed, and there's a fine line between giving a credit for original work and impersonating someone with changes they might not even do, so figure I'd ask explicitly about this matter ;-) ).

pfalcon commented:
@jhedberg was added to this review automatically because the PR contained earlier version of https://github.com/zephyrproject-rtos/zephyr/pull/12230 . It was merged and I rebased this PR, so removing extra reviewers to keep everyone's review list clean. (Of course, reviews are always welcome.)

galak commented:
> @galak, let me know if you agree with the proposal in p.6 above. And let me know if you care if that commit is under your authorship or care it to be now (it would be very different from what _you_ committed, and there's a fine line between giving a credit for original work and impersonating someone with changes they might not even do, so figure I'd ask explicitly about this matter ;-) ).

I'm good w/6.  The pulling from mbedOS was an attempt to make things quicker to get something working.  I don't think we need to maintain any linage to it.

pfalcon committed:
drivers: eth_smsc911x: Add driver for SMSC9118 aka LAN9118 chip

As emulated by QEMU. SMSC9118 is compatible with SMSC9220 as used in
ARM MPS2 board, as well as SMSC9115/6/7/etc. devices.

Portions of the code are based on mbedOS code from its
targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/device/drivers/smsc9220_eth.c

eth_smsc9220_priv.h originally comes from Arm mbedOS file:

targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/device/drivers/smsc9220_eth.h

augmented with struct & defines from:

targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/device/CM3DS.h

and renamed as eth_smsc911x_priv.h to follow Zephyr conventions.

Then, following changes applied:

Changes to build under Zephyr, changes to use symbolic constants
and field access helpers, typo fixes, etc.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
@galak:

> I'm good w/6.

Thanks, refactored, commits look much cleaner now.

pfalcon committed:
samples: sockets: Add eth_smsc911x overlays

Add overlays for commonly used socket samples: big_http_download and
dumb_http_server. Can be used with:

make BOARD=mps2_an385 CONF_FILE="prj.conf overlay-smsc911x.conf"

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Otherwise, the only remaining issue here is logging from ISR. With current logging, it just crashes with assert. I have high hopes that https://github.com/zephyrproject-rtos/zephyr/pull/12406 might do something about that (TODO to review/test it).

So, I would welcome reviews coming in now.

pfalcon commented:
Ok, logging issues were due to enabling CONFIG_LOG_INPLACE_PROCESS, and now ticketed as https://github.com/zephyrproject-rtos/zephyr/issues/12494 . Logging workarounds were removed from this PR now, and it contains only the relevant stuff.

@galak, @jukkar, @tbursztyka : Please review.

tbursztyka commented on drivers/ethernet/CMakeLists.txt:
```diff
@@ -10,6 +10,7 @@ zephyr_sources_ifdef(CONFIG_ETH_DW		eth_dw.c)
 zephyr_sources_ifdef(CONFIG_ETH_E1000		eth_e1000.c)
 zephyr_sources_ifdef(CONFIG_ETH_ENC28J60	eth_enc28j60.c)
 zephyr_sources_ifdef(CONFIG_ETH_MCUX		eth_mcux.c)
+zephyr_sources_ifdef(CONFIG_ETH_SMSC911X	eth_smsc911x.c)
```
Merge this patch with previous one. Driver comes all file at once: code, Kconfig and built directives

pfalcon replied:
Hmm, didn't notice we had it like that. Done.

tbursztyka commented on drivers/ethernet/eth_smsc911x_priv.h:
```diff
@@ -0,0 +1,214 @@
+/* mbed Microcontroller Library
```
Merge this with next patch: driver code should come all at once.

pfalcon replied:
Done.

tbursztyka commented on drivers/ethernet/eth_smsc911x.c:
```diff
@@ -0,0 +1,718 @@
+	if (smsc_phy_regread(SMSC9220_PHY_ID1, &phyid1)) {
+		return -1;
+	}
+	if (smsc_phy_regread(SMSC9220_PHY_ID2, &phyid2)) {
```
have an empty line after the }, same below at line 262

pfalcon replied:
Done.

tbursztyka commented on drivers/ethernet/eth_smsc911x.c:
```diff
@@ -0,0 +1,718 @@
+	}
+
+	val |= 1 << 15;
+
```
you can remove that line since you use val in the if ()

pfalcon replied:
> you can remove that line since you use val in the if ()

Well, I'd say it's more symmetrical (read-modify-write) and easier to read the original way.

tbursztyka commented on drivers/ethernet/eth_smsc911x.c:
```diff
@@ -0,0 +1,718 @@
+	/* threshold to defaults specified. */
+	SMSC9220->AFC_CFG = 0x006E3740;
+
+#if 0
```
Get rid of this code if it's never meant to be compiled.

pfalcon replied:
> Get rid of this code if it's never meant to be compiled.

So, this driver is written for hardware which actually exists, even though it's probably will be mostly used for QEMU emulated (== simplified) hardware, and that's how I tested it (I don't have access to MPS2 board). So, I wouldn't like to cut (or complicate) way to make it run well on real hw too. The original mbedOS driver had this code, I didn't convert it as QEMU doesn't emulate EEPROM, but whenever possible, I'd prefer to leave it in the driver even if commented out.  I added comment why it is so.

jukkar replied:
> So, I wouldn't like to cut (or complicate) way to make it run well on real hw too

Please do it then like this:
```
	/* Not needed on QEMU, but may be needed on real hardware. */
	if (!IS_ENABLED(CONFIG_QEMU_TARGET) && smsc_wait_eeprom()) {
		return -1;
	}
```

pfalcon replied:
> Please do it then like this:
>	if (!IS_ENABLED(CONFIG_QEMU_TARGET) && smsc_wait_eeprom()) 

That's confusing and misleading: not reading EEPROM doesn't depend on being QEMU_TARGET;  smsc_wait_eeprom() doesn't exit.

It's just a TODO item/note to someone else who may want to get that driver running on the real HW, it should be good as it is.

jukkar replied:
> That's confusing and misleading: not reading EEPROM doesn't depend on being QEMU_TARGET; smsc_wait_eeprom() doesn't exit.

Your comment is then misleading as it says that calling smsc_wait_eeprom() is not needed on QEMU. Anyway, lets remove this piece of code that is never going to be called, it is just confusing everybody. If we run this with real hardware, then suitable spells will be added here anyway.

pfalcon replied:
> Anyway, lets remove this piece of code

Ok, done.

tbursztyka commented on drivers/ethernet/eth_smsc911x.c:
```diff
@@ -0,0 +1,718 @@
+	u32_t txcmd_a, txcmd_b;
+	u32_t tx_stat;
+
+	txcmd_a = (1/*is_first_segment*/ << 13) | (1/*is_last_segment*/ << 12)
```
You could have created dedicated macros for these bits.

pfalcon replied:
> You could have created dedicated macros for these bits.

I could, and actually I converted quite a bunch of raw numbers like that as they were in the original mbedOS. But not everything, and as this PR already spent a lot of time in works, with a lot of side-tracking, I kinda like it to pronounced done finally ;-)

tbursztyka commented on drivers/ethernet/eth_smsc911x.c:
```diff
@@ -0,0 +1,718 @@
+
+static struct net_pkt *smsc_recv_pkt(u32_t pkt_size)
+{
+		u32_t rem_size;
```
all the code in this function seems be intended one tab too many

pfalcon replied:
> all the code in this function seems be intended one tab too many

Indeed, fixed.

tbursztyka reviewed (changes requested):
Minor comments, overall seems fine to me.

pfalcon commented:
Updated addressing @tbursztyka's comments.

nashif reviewed (changes requested):
Can we add this to the yaml file so it actually builds for this sample?

pfalcon commented:
> Can we add this to the yaml file so it actually builds for this sample?

@nashif, What exactly do you mean?

pfalcon commented:
> Can we add this to the yaml file so it actually builds for this sample?

Ok, I assume sample.yaml echo_server, which already has a bunch of overlay tests. Doing.

pfalcon committed:
samples: echo_server: Test overlay-smsc911x.conf via sanitycheck

By adding relevant test config to sample.yaml.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Ok, added sample.yaml change requested by @nashif and rebased on master. CI passes.

@tbursztyka, thanks for review, @jukkar, please have a look too.

jukkar commented on drivers/ethernet/eth_smsc911x.c:
```diff
@@ -0,0 +1,721 @@
+{
+	ARG_UNUSED(dev);
+
+	return ETHERNET_LINK_10BASE_T | ETHERNET_LINK_100BASE_T;
```
Please consider to add VLAN support later. This is not preventing merging now but it would be good to have that supported.

jukkar commented on drivers/ethernet/eth_smsc911x.c:
```diff
@@ -0,0 +1,721 @@
+	struct device *dev = net_if_get_device(iface);
+	struct eth_context *context = dev->driver_data;
+	/* Must be static, netif has pointer stored */
+	static u8_t mac[6];
```
Why not add the mac string to eth_context instead?

pfalcon replied:
> Why not add the mac string to eth_context instead?

Done

jukkar commented on drivers/ethernet/eth_smsc911x.c:
```diff
@@ -0,0 +1,721 @@
+	int res;
+	u16_t total_len = net_pkt_get_len(pkt);
+	u16_t sent = 0;
+	static u8_t tx_buf[1540] __aligned(4);
```
Why 1540 bytes here? For example in mcux, it allocates 1514 bytes buffer (max mtu + ethernet header size)

pfalcon replied:
> Why 1540 bytes here? For example in mcux, it allocates 1514

Fixed.

jukkar commented on drivers/ethernet/eth_smsc911x.c:
```diff
@@ -0,0 +1,721 @@
+	SMSC9220->TX_DATA_PORT = txcmd_a;
+	SMSC9220->TX_DATA_PORT = txcmd_b;
+
+#define LINEARIZE
```
Why not have this as a Kconfig option, or just remove the unused branch if it is not working according to your comment.

pfalcon replied:
> Why not have this as a Kconfig option, or just remove the unused branch if it is not working according to your comment.

Definitely not ready to be a Kconfig option. I just implemented it initially w/o intermediate linearization step, but it got broken when @tbursztyka introduced arbitrary-sized fragments support, and I was reluctant to remove the original code, but did that now.

jukkar reviewed (changes requested):
Looks good, couple of minor nitpicks

pfalcon commented:
@jukkar: Please re-review.
